Sources/NIOSSL/SSLCertificate.swift

```diff
@@ -284,6 +284,44 @@ extension NIOSSLCertificate: Equatable {
     }
 }
 
+
+extension NIOSSLCertificate: Hashable {
+    /// This is a silly thunk type that exists just to provide stack-allocation for a SHA256 digest.
+    private struct SHA256Digest {
+        // A SHA256 digest is 256 bits long, or 32 bytes, or 4 UInt64s. We store this
+        // as UInt64s here, but they are always accessed as UInt8s.
+        private var storage: (UInt64, UInt64, UInt64, UInt64)
+
+        init() {
+            self.storage = (0, 0, 0, 0)
+        }
+
+        mutating func withUnsafeMutablePointer<T>(_ body: (UnsafeMutableBufferPointer<UInt8>) throws -> T) rethrows -> T {
+            return try Swift.withUnsafeMutablePointer(to: &self.storage) { (storagePtr: UnsafeMutablePointer<(UInt64, UInt64, UInt64, UInt64)>) -> T in
+                return try storagePtr.withMemoryRebound(to: UInt8.self, capacity: 1) { bytesPtr -> T in
+                    let buffer = UnsafeMutableBufferPointer(start: bytesPtr, count: MemoryLayout<(UInt64, UInt64, UInt64, UInt64)>.size)
+                    return try body(buffer)
+                }
+            }
+        }
+    }
+
+
+    public func hash(into hasher: inout Hasher) {
+        // Conveniently BoringSSL gives us a useful function for this: we can ask it to provide a digest of the X509 object.
+        // That's what we're going to do here. We use SHA256 here because it's secure and fast.
+        var digest = SHA256Digest()
+
+        digest.withUnsafeMutablePointer {
+            var length = CUnsignedInt(0)
+            let rc = CNIOBoringSSL_X509_digest(self.ref, CNIOBoringSSL_EVP_sha256(), $0.baseAddress, &length)
```

Comment:
I think conversion to ASN.1 DER with a subsequent `hasher(bytes: )` is more efficient than this. The X509_digest does that conversion into DER (with a corresponding memory allocation) internally, then we're applying SHA256 on top, where for all practical purposes a simpler byte hashing would be significantly faster than properly digesting using a cryptographically sound hash.

Just a nit.

Reply:
Yeah, I had wondered about making that change. It’s always more annoying to have to write more code though. :wink: I’ll sit with the idea this weekend and decide how much more time I want to spend on the process.